contracts/PSM/BasePSM.sol

```diff
@@ -0,0 +1,135 @@
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Ownable2Step } from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import { Fixed256x18 } from "@tempusfinance/tempus-utils/contracts/math/Fixed256x18.sol";
+import { ERC20PermitSignature } from "@tempusfinance/tempus-utils/contracts/utils/PermitHelper.sol";
+import { IRToken } from "../Interfaces/IRToken.sol";
+import { IPositionManager } from "../Interfaces/IPositionManager.sol";
+import { IPSM } from "./IPSM.sol";
+import { IPSMFeeCalculator } from "./IPSMFeeCalculator.sol";
+
+/// @dev Base implementation of Peg Stability Module.
+/// Handles basic reserve token transfers as well as minting R.
+abstract contract BasePSM is IPSM, ERC20, Ownable2Step {
+    using SafeERC20 for IERC20;
+    using Fixed256x18 for uint256;
+
+    IERC20 public immutable override reserveToken;
+    IRToken public immutable override rToken;
+    IPositionManager public immutable override positionManager;
+    IPSMFeeCalculator public override feeCalculator;
+
+    constructor(
+        IERC20 reserveToken_,
+        IRToken rToken_,
+        string memory name_,
+        string memory symbol_,
+        IPSMFeeCalculator feeCalculator_
+    )
+        ERC20(name_, symbol_)
+    {
+        if (address(reserveToken_) == address(0)) {
+            revert ZeroInputProvided();
+        }
+
+        reserveToken = reserveToken_;
+        rToken = rToken_;
+        positionManager = IPositionManager(rToken_.positionManager());
+
+        setFeeCalculator(feeCalculator_);
+
+        _approve(address(this), address(positionManager), type(uint256).max);
+    }
+
+    function buyR(uint256 reserveAmount, uint256 minReturn) external {
+        if (reserveAmount == 0 || minReturn == 0) {
+            revert ZeroInputProvided();
+        }
+
+        reserveToken.safeTransferFrom(msg.sender, address(this), reserveAmount);
+        _depositReserveToken(reserveAmount);
+
+        uint256 mintAmount = reserveAmount - feeCalculator.calculateFee(reserveAmount, true);
+        if (mintAmount < minReturn) {
+            revert ReturnLessThanMinimum(mintAmount, minReturn);
+        }
+        _mintR(msg.sender, mintAmount);
+    }
+
+    function buyReserveToken(uint256 rAmount, uint256 minReturn) external {
+        if (rAmount == 0 || minReturn == 0) {
+            revert ZeroInputProvided();
+        }
+
+        _burnR(msg.sender, rAmount);
+
+        uint256 amountMinusFee = rAmount - feeCalculator.calculateFee(rAmount, false);
+        if (amountMinusFee < minReturn) {
+            revert ReturnLessThanMinimum(amountMinusFee, minReturn);
+        }
+        _withdrawReserveToken(amountMinusFee);
+        reserveToken.safeTransfer(msg.sender, amountMinusFee);
+    }
+
+    function withdrawReserve(address to, uint256 amount) external override onlyOwner {
```

Comment:
maybe let's just add a generic `recover` function?

Reply:
I would add `recover` on top of the current one. There might be some reserve integration that is not erc20 :)